worktree_status.go

```diff
@@ -243,30 +251,147 @@ func diffTreeIsEquals(a, b noder.Hasher) bool {
 }
 
 // Add adds the file contents of a file in the worktree to the index. if the
-// file is already staged in the index no error is returned.
+// file is already staged in the index no error is returned. If a file deleted
+// from the Workspace is given, the file is removed from the index. If a
+// directory given, adds the files and all his sub-directories recursively in
+// the worktree to the index. If any of the files is already staged in the index
+// no error is returned. When path is a file, the blob.Hash is returned.
 func (w *Worktree) Add(path string) (plumbing.Hash, error) {
+	// TODO(mcuadros): remove plumbing.Hash from signature at v5.
 	s, err := w.Status()
 	if err != nil {
 		return plumbing.ZeroHash, err
 	}
 
-	h, err := w.copyFileToStorage(path)
+	idx, err := w.r.Storer.Index()
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	var h plumbing.Hash
+	var added bool
+
+	fi, err := w.Filesystem.Lstat(path)
+	if err != nil || !fi.IsDir() {
+		added, h, err = w.doAddFile(idx, s, path)
+	} else {
+		added, err = w.doAddDirectory(idx, s, path)
+	}
+
 	if err != nil {
-		if os.IsNotExist(err) {
-			h, err = w.deleteFromIndex(path)
-		}
 		return h, err
 	}
 
-	if s.File(path).Worktree == Unmodified {
+	if !added {
 		return h, nil
 	}
 
-	if err := w.addOrUpdateFileToIndex(path, h); err != nil {
-		return h, err
+	return h, w.r.Storer.SetIndex(idx)
+}
+
+func (w *Worktree) doAddDirectory(idx *index.Index, s Status, directory string) (added bool, err error) {
+	files, err := w.Filesystem.ReadDir(directory)
+	if err != nil {
+		return false, err
+	}
+
+	for _, file := range files {
+		name := path.Join(directory, file.Name())
+
+		var a bool
+		if file.IsDir() {
+			a, err = w.doAddDirectory(idx, s, name)
+		} else {
+			a, _, err = w.doAddFile(idx, s, name)
+		}
+
+		if err != nil {
+			return
+		}
+
+		if !added && a {
+			added = true
+		}
+	}
+
+	return
+}
+
+// AddGlob adds all paths, matching pattern, to the index. If pattern matches a
+// directory path, all directory contents are added to the index recursively. No
+// error is returned if all matching paths are already staged in index.
+func (w *Worktree) AddGlob(pattern string) error {
+	files, err := util.Glob(w.Filesystem, pattern)
+	if err != nil {
+		return err
+	}
+
+	if len(files) == 0 {
+		return ErrGlobNoMatches
+	}
+
+	s, err := w.Status()
+	if err != nil {
+		return err
+	}
+
+	idx, err := w.r.Storer.Index()
+	if err != nil {
+		return err
+	}
+
+	var saveIndex bool
+	for _, file := range files {
```

Comment:
This also looks like we aren't trying to match patterns on things in the index, only paths that exist in the working directory. This is slightly different behavior than `git add -- <glob>` in cases where index entries should be removed (i.e. `git add` supports both adding AND removing entries from the index).

Reply:
Add, only adds files from the woktree to the index. The function `Add` should not do exactly the behavior of `git add`. This is a programmatically interface, so I rather keep it simple.

Reply:
I think that makes sense. I'm just anticipating that this discrepancy will cause confusion for users and drive more issues being opened.